src/Orleans/Messaging/SocketManager.cs

```diff
@@ -65,9 +67,16 @@ private Socket SendingSocketCreator(IPEndPoint target)
                 s.NoDelay = true;
                 WriteConnectionPreemble(s, Constants.SiloDirectConnectionId); // Identifies this client as a direct silo-to-silo socket
                 // Start an asynch receive off of the socket to detect closure
-                var foo = new byte[4];
-                s.BeginReceive(foo, 0, 1, SocketFlags.None, ReceiveCallback,
-                    new Tuple<Socket, IPEndPoint, SocketManager>(s, target, this));
+                var receiveAsyncEventArgs = new SocketAsyncEventArgs
+                {
+                    BufferList = new List<ArraySegment<byte>> { new ArraySegment<byte>(new byte[4]) },
+                    UserToken = new Tuple<Socket, IPEndPoint, SocketManager>(s, target, this)
+                };
+                receiveAsyncEventArgs.Completed += ReceiveCallback;
+                if (!s.ReceiveAsync(receiveAsyncEventArgs))
```

Comment:
@veikkoeeva I'm not at all concerned about perf, since as it's commented there, we do not expect to ever complete that receive, until the connection is shutting down. I just changed to use an API that is supported.
I'm still interested to know if this is reliable though, as what I tried to do here is refactor without external side effects. Hopefully I achieved that.
Let me know if there's something I missed here that concerns you with regards to perf though.